### ISSUE TITLE
module 'torch.mps' has no attribute 'device'

### ISSUE THREAD

**Issue by david-waterworth** [pytorch/pytorch#127676]
### 🐛 Describe the bug

I'm getting the error `module 'torch.mps' has no attribute 'device'` trying to train a `sentence-transformer model (which uses the huggingface transformers trainer).

The error is raised on this line -> https://github.com/pytorch/pytorch/blob/7ef7c265d4361691dc4cf54152db083de3215fbf/torch/utils/checkpoint.py#L183

Because the mps model doesn't have a `device()` method

https://github.com/pytorch/pytorch/blob/7ef7c265d4361691dc4cf54152db083de3215fbf/torch/mps/__init__.py

### Versions

Collecting environment information...
PyTorch version: 2.3.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 14.5 (arm64)
GCC version: Could not collect
Clang version: 15.0.0 (clang-1500.3.9.4)
CMake version: Could not collect
Libc version: N/A

Python version: 3.10.14 (main, May 25 2024, 13:44:20) [Clang 15.0.0 (clang-1500.3.9.4)] (64-bit runtime)
Python platform: macOS-14.5-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M3 Pro

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] torch==2.3.0
[conda] Could not collect

**Comment by tedunderwood** [comment]
I am getting the same bug under Python 3.13 and OS 14.3.1. I got it under both torch 2.2.2 and torch 2.3.

**Comment by albanD** [comment]
Hey!

I think there are two ways we can go about this one:
- Update the mps device_module to add this function to set the current device.
- Change the checkpoint code to only call this function if the current device is not the same as the device we want to set (so that device/accelerator that only ever have one device don't need to worry about it).

**Comment by tedunderwood** [comment]
For what it's worth, I only encounter this error when using CachedGISTEmnbeddingLoss or CachedMultipleNegativesRankingLoss. The non-cached versions of those same loss functions do not create the error.

**Comment by kulinseth** [comment]
> Hey!
> 
> I think there are two ways we can go about this one:
> 
> * Update the mps device_module to add this function to set the current device.
> * Change the checkpoint code to only call this function if the current device is not the same as the device we want to set (so that device/accelerator that only ever have one device don't need to worry about it).

I like option 1. We would need this method later for Triton work.. Updating [current PR](https://github.com/pytorch/pytorch/pull/128289) with it.
